summoner-cli/src/Summoner/Template/GitHub.hs

```diff
@@ -121,49 +121,121 @@ gitHubFiles Settings{..} = concat
         , "        cabal-version: ${{ matrix.cabal }}"
         , ""
-        , "    - uses: actions/cache@v1"
+        , "    - name: Freeze"
+        , "      run: |"
+        , "        cabal v2-freeze"
+        , ""
+        , "    - uses: actions/cache" <> ghcActionsCacheVersion
         , "      name: Cache ~/.cabal/store"
         , "      with:"
-        , "        path: ~/.cabal/store"
-        , "        key: ${{ runner.os }}-${{ matrix.ghc }}-cabal"
+        , "        path: ${{ steps.setup-haskell-cabal.outputs.cabal-store }}"
+        , "        key: ${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('cabal.project.freeze') }}"
         , ""
         , "    - name: Build"
         , "      run: |"
-        , "        " <> cabalUpdate
+        , "        " <> cabalConfigure
         , "        " <> cabalBuild
         , ""
         , "    - name: Test"
         , "      run: |"
         , "        " <> cabalTest
+        , ""
         ]
 
+    ghActionsStack :: [Text]
+    ghActionsStack =
+        [ "  stack:"
+        , "    name: stack / ghc ${{ matrix.ghc }}"
+        , "    runs-on: ubuntu-latest"
+        , "    strategy:"
+        , "      matrix:"
+        , "        stack: [" <> quote defaultStack <> "]"
+        , "        ghc:"
+        ]
+        <> map (indent 10 <>) ghActionsVersions
+        <>
+        [ ""
+        , "    steps:"
+        , "    - uses: actions/checkout" <> ghcActionsCheckoutVersion
+        , "      if: github.event.action == 'opened' || github.event.action == 'synchronize' || github.event.ref == 'refs/heads/master'"
+        , ""
+        , "    - uses: actions/setup-haskell" <> ghcActionsSetupHaskellVersion
+        , "      name: Setup Haskell Stack"
+        , "      with:"
+        , "        ghc-version: ${{ matrix.ghc }}"
+        , "        stack-version: ${{ matrix.stack }}"
+        , ""
+        , "    - uses: actions/cache" <> ghcActionsCacheVersion
+        , "      name: Cache ~/.stack"
+        , "      with:"
+        , "        path: ~/.stack"
+        , "        key: ${{ runner.os }}-${{ matrix.ghc }}-stack"
+        , ""
+        , "    - name: Build"
+        , "      run: |"
+        , "        " <> stackBuild
+        , ""
+        , "    - name: Test"
+        , "      run: |"
+        , "        " <> stackTest
+        ]
 
     ghActionsVersions :: [Text]
     ghActionsVersions = map
         (\ghc -> "- " <> quote (showGhcVer ghc))
         settingsTestedVersions
 
+    ghActionsExcludes :: [Text]
+    ghActionsExcludes =
+        let latestVersion = viaNonEmpty last $ sort settingsTestedVersions
+            versionsToExclude = filter (\version -> pure version /= latestVersion) settingsTestedVersions
+            osesToExclude = ["macOS-latest", "windows-latest"]
+            excludes = do
+                os <- osesToExclude
+                version <- versionsToExclude
+                [  "- os: " <> os
+                 , "  ghc: " <> showGhcVer version
+                 ]
```

Comment:
Very clever implementation! 👍🏼 🧠

Reply:
Thanks :) If in this case clever goes with hard to read then I'm willing to replace. Two alternatives that I considered are list comprehension (I see the benefit of being able to merge the filter part as a guard) and dropping `do` notation.